From 39963226535a7b13823902c22dcf399c27965923 Mon Sep 17 00:00:00 2001
From: Zhikui Ren <zhikui.ren@intel.com>
Date: Thu, 4 May 2023 17:06:50 -0700
Subject: [PATCH 1/7] fansensor: fix fan sensor service crash

fs::canonical can throw exception if the path does not exist,
which can happen for i2c fan type. Check for path exists before the
call.

Tested:
fan sensor service successfully detect i2c fan type and does not crash.

Fixes: 9a472e8 ("detect fan type by matching against 'compatible'")
Signed-off-by: Zhikui Ren <zhikui.ren@intel.com>
Change-Id: I395e4d47b16fc58eba777e88be304bf28993bec3
---
 src/FanMain.cpp | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/src/FanMain.cpp b/src/FanMain.cpp
index 32022c1..85de747 100644
--- a/src/FanMain.cpp
+++ b/src/FanMain.cpp
@@ -74,8 +74,12 @@ static const std::map<std::string, FanTypes> compatibleFanTypes = {
 FanTypes getFanType(const fs::path& parentPath)
 {
     fs::path linkPath = parentPath / "of_node";
-    std::string canonical = fs::canonical(linkPath);
+    if (!fs::exists(linkPath))
+    {
+        return FanTypes::i2c;
+    }
 
+    std::string canonical = fs::canonical(linkPath);
     std::string compatiblePath = canonical + "/compatible";
     std::ifstream compatibleStream(compatiblePath);
 
-- 
2.39.2

